widget/list.go

```diff
@@ -0,0 +1,415 @@
+	if size != l.size {
+		if size.Width != l.size.Width {
+			for _, child := range l.children {
+				child.Resize(fyne.NewSize(size.Width, l.list.itemMin.Height))
+			}
+		}
+		l.scroller.Resize(size)
+		l.size = size
+	}
+	if l.list.Length() == 0 {
+		return
+	}
+	if l.itemCache == nil {
+		l.itemCache = &listPool{}
+	}
+
+	offsetChange := l.list.previousOffsetY - l.list.offsetY
+
+	if offsetChange < 0 {
+		// Scrolling Down
+		itemChange := 0
+		layoutEndY := l.children[len(l.children)-1].Position().Y + l.list.itemMin.Height
+		scrollerEndY := l.scroller.Offset.Y + l.scroller.Size().Height
+		if layoutEndY < scrollerEndY {
+			itemChange = int(math.Ceil(float64(scrollerEndY-layoutEndY) / float64(l.list.itemMin.Height)))
+		} else if offsetChange*-1 < l.list.itemMin.Height {
+			return
+		} else {
+			itemChange = int(math.Floor(float64(offsetChange*-1) / float64(l.list.itemMin.Height)))
+		}
+		l.list.previousOffsetY = l.list.offsetY
+		for i := 0; i < itemChange && l.lastItemIndex != l.list.Length()-1; i++ {
+			l.itemCache.Release(l.children[0])
+			l.children = l.children[1:]
+			l.firstItemIndex++
+			l.lastItemIndex++
+			l.appendItem(l.lastItemIndex)
+		}
+	} else if offsetChange > 0 {
+		// Scrolling Up
+		itemChange := 0
+		layoutStartY := l.children[0].Position().Y
+		if layoutStartY > l.scroller.Offset.Y {
+			itemChange = int(math.Ceil(float64(layoutStartY-l.scroller.Offset.Y) / float64(l.list.itemMin.Height)))
+		} else if offsetChange < l.list.itemMin.Height {
+			return
+		} else {
+			itemChange = int(math.Floor(float64(offsetChange) / float64(l.list.itemMin.Height)))
+		}
+		l.list.previousOffsetY = l.list.offsetY
+		for i := 0; i < itemChange && l.firstItemIndex != 0; i++ {
+			l.itemCache.Release(l.children[len(l.children)-1])
+			l.children = l.children[:len(l.children)-1]
+			l.firstItemIndex--
+			l.lastItemIndex--
+			l.prependItem(l.firstItemIndex)
+		}
+	} else {
+		// Relayout What Is Visible - no scroll change - initial layout or possibly from a resize
+		l.visibleItemCount = int(math.Ceil(float64(l.scroller.size.Height) / float64(l.list.itemMin.Height)))
+		if len(l.children) >= l.visibleItemCount || len(l.children) >= l.list.Length() {
+			return
+		}
+		for i := len(l.children) + l.firstItemIndex; len(l.children) <= l.visibleItemCount && i < l.list.Length(); i++ {
+			l.appendItem(i)
+		}
+		l.layout.Objects = l.children
+		l.layout.Layout.Layout(l.layout.Objects, l.list.itemMin)
+		l.lastItemIndex = l.firstItemIndex + len(l.children) - 1
+
+		poolCapacity := 5
+		if l.list.Length()-len(l.children) < 5 {
+			poolCapacity = l.list.Length() - len(l.children)
+		}
+		for i := l.itemCache.Count(); i < poolCapacity; i++ {
+			// Make sure to keep extra items in the cache
+			item := newListItem(l.list.CreateItem(), nil)
+			l.itemCache.Release(item)
+		}
+	}
+}
+
+func (l *listRenderer) MinSize() fyne.Size {
+	return l.scroller.MinSize()
+}
+
+func (l *listRenderer) Refresh() {
+	l.Layout(l.list.Size())
```

Comment:
I'm not sure why the refresh causes a layout here - it may pair with the "size" cache that was added above.

Reply:
because it refreshes on scroll change and calls layout